pkg/controller/proxyconfig/validation.go

```diff
@@ -52,5 +56,87 @@ func (r *ReconcileProxyConfig) ValidateProxyConfig(proxyConfig *configv1.ProxySp
 		}
 	}
 
+	if isSpecReadinessEndpoints(proxyConfig) {
+		for _, endpoint := range proxyConfig.ReadinessEndpoints {
+			scheme, err := validation.URI(endpoint)
+			if err != nil {
+				return fmt.Errorf("invalid URI for endpoint %s: %v", endpoint, err)
+			}
+			switch {
+			case scheme == proxyHTTPScheme:
+				// TODO: Add case for proxyHTTPSScheme when CA support is merged.
+				if err := validateHTTPReadinessEndpoint(proxyConfig.HTTPProxy, endpoint); err != nil {
+					return fmt.Errorf("readinessEndpoint probe failed for endpoint %s", endpoint)
+				}
+			default:
+				// TODO: Update error to include proxyHTTPSScheme when CA support is merged.
+				return fmt.Errorf("readiness endpoints requires a %q URI sheme", proxyHTTPScheme)
+			}
+		}
+	}
+
 	return nil
 }
+
+// validateHTTPReadinessEndpoint validates an http readinessEndpoint endpoint.
+func validateHTTPReadinessEndpoint(httpProxy, endpoint string) error {
+	if err := validateHTTPReadinessEndpointWithRetries(httpProxy, endpoint, proxyProbeMaxRetries); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateHTTPReadinessEndpointWithRetries tries to validate an http
+// endpoint in a finite loop and returns the last result if it never succeeds.
+func validateHTTPReadinessEndpointWithRetries(httpProxy, endpoint string, retries int) error {
+	var err error
+	for i := 0; i < retries; i++ {
+		err = runHTTPReadinessProbe(httpProxy, endpoint)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+// runHTTPReadinessProbe issues an http GET request to endpoint and returns
+// an error if a 2XX or 3XX http status code is not returned. The request
+// is proxied if proxy environment variables exist.
+func runHTTPReadinessProbe(httpProxy, endpoint string) error {
+	// TODO: Uncomment when proxy test environment is available to
+	//  force transport to use a proxy.
+	/*proxyURL, err := url.Parse(httpProxy)
+	if err != nil {
+		return fmt.Errorf("failed to parse httpProxy url '%s': %v", httpProxy, err)
+	}*/
+
+	transport := &http.Transport{
+		// TODO: Uncomment when proxy test environment is available to
+		//  force transport to use a proxy.
+		//Proxy: http.ProxyURL(proxyURL),
+		Proxy: http.ProxyFromEnvironment,
```

Comment:
you shouldn't be getting the proxy from the environment, you should be using the proxy from proxyconfig.spec since that's the proxy you want to make sure actually works.  We want to validate the spec value because that is the one the administrator is asserting the cluster should start using, and we want to confirm it works before we propagate it into status (at which point all the consumers, including the network operator itself, will pick it up and start using it).

Reply:
@bparees Understood, see the lines that I commented out. `proxyURL` is the parsed URL from proxy.spec.httpProxy that gets passed in as `httpProxy`. My plan was to change this after we have a working CI. I don't have a proxy setup in my dev env. Maybe I should run a local proxy since I'm running CNO locally for testing.

Reply:
I updated the http probe transport to use proxy from `proxy.spec.httpProxy` instead of from env var.